In [1]:
# Get imports
import pandas as pd
import numpy as np 

In [2]:
# Let's make a function that takes in a database name as a string
# this function also performs our imports from env
def get_db_url(db_name):
    from env import user, host, password
    return f'mysql+pymysql://{user}:{password}@{host}/{db_name}'

In [3]:
sql = """
    SELECT * from employees limit 3000
"""

# 5
Once you have successfully run a query:

a. Intentionally make a typo in the database url. What kind of error message do you see?

    - NoSuchModuleError: Can't load plugin: sqlalchemy.dialects:amysql.pymysql

b. Intentionally make an error in your SQL query. What does the error message look like?

    - ProgrammingError: (pymysql.err.ProgrammingError) (1064, "You have an error in your SQL syntax; check the manual that corresponds to your MySQL server version for the right syntax to use near 'fro employees limit 3000' at line 1")
[SQL: 
    SELECT * fro employees limit 3000
]
(Background on this error at: http://sqlalche.me/e/14/f405)

# 6 
Read the employees and titles tables into two separate DataFrames.

In [7]:
query_6_1 = '''
SELECT * 
FROM employees
'''

query_6_2 = '''
SELECT * 
FROM titles
'''

In [8]:
url = get_db_url("employees")
df_6_employees = pd.read_sql(query_6_1, url)
df_6_employees

emp_no  birth_date first_name last_name gender   hire_date
0        10001  1953-09-02     Georgi   Facello      M  1986-06-26
1        10002  1964-06-02    Bezalel    Simmel      F  1985-11-21
2        10003  1959-12-03      Parto   Bamford      M  1986-08-28
3        10004  1954-05-01  Chirstian   Koblick      M  1986-12-01
4        10005  1955-01-21    Kyoichi  Maliniak      M  1989-09-12
...        ...         ...        ...       ...    ...         ...
300019  499995  1958-09-24     Dekang  Lichtner      F  1993-01-12
300020  499996  1953-03-07       Zito      Baaz      M  1990-09-27
300021  499997  1961-08-03    Berhard    Lenart      M  1986-04-21
300022  499998  1956-09-05   Patricia   Breugel      M  1993-10-13
300023  499999  1958-05-01     Sachin   Tsukuda      M  1997-11-30

[300024 rows x 6 columns]

In [9]:
df_6_titles = pd.read_sql(query_6_2, url)
df_6_titles

emp_no            title   from_date     to_date
0        10001  Senior Engineer  1986-06-26  9999-01-01
1        10002            Staff  1996-08-03  9999-01-01
2        10003  Senior Engineer  1995-12-03  9999-01-01
3        10004         Engineer  1986-12-01  1995-12-01
4        10004  Senior Engineer  1995-12-01  9999-01-01
...        ...              ...         ...         ...
443303  499997         Engineer  1987-08-30  1992-08-29
443304  499997  Senior Engineer  1992-08-29  9999-01-01
443305  499998     Senior Staff  1998-12-27  9999-01-01
443306  499998            Staff  1993-12-27  1998-12-27
443307  499999         Engineer  1997-11-30  9999-01-01

[443308 rows x 4 columns]

# 7 
How many rows and columns do you have in each DataFrame? Is that what you expected?

    - Employees 300,0024 rows x 6 columns
    - Titles 443,308 rows x 4 columns 
    Yes, I was expecting the titles table to have more records because one person can habe multiple jobs 
    



# 8 
Display the summary statistics for each DataFrame

In [22]:
df_6_employees.describe()

emp_no
count  300024.000000
mean   253321.763392
std    161828.235540
min     10001.000000
25%     85006.750000
50%    249987.500000
75%    424993.250000
max    499999.000000

In [13]:
df_6_titles.describe()

emp_no
count  443308.000000
mean   253075.034430
std    161853.292613
min     10001.000000
25%     84855.750000
50%    249847.500000
75%    424891.250000
max    499999.000000

# 9 
How many unique titles are in the titles DataFrame?

In [34]:
df_6_titles.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 443308 entries, 0 to 443307
Data columns (total 4 columns):
 #   Column     Non-Null Count   Dtype 
---  ------     --------------   ----- 
 0   emp_no     443308 non-null  int64 
 1   title      443308 non-null  object
 2   from_date  443308 non-null  object
 3   to_date    443308 non-null  object
dtypes: int64(1), object(3)
memory usage: 13.5+ MB


In [19]:
df_6_titles.title.unique()

array(['Senior Engineer', 'Staff', 'Engineer', 'Senior Staff',
       'Assistant Engineer', 'Technique Leader', 'Manager'], dtype=object)

# 10
What is the oldest date in the to_date column?

In [29]:
df_6_titles.to_date.min()

datetime.date(1985, 3, 1)

# 11
What is the most recent date in the to_date column?


In [33]:
df_6_titles.to_date.max()

datetime.date(9999, 1, 1)